src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs

```diff
@@ -561,43 +562,69 @@ private void OnDependencyPropertyChanged(DependencyPropertyDetails propertyDetai
 		{
 			SetBindingValue(propertyDetails, args.NewValue);
 
-			var (hasValueInherits, hasValueDoesNotInherit) = GetPropertyInheritanceConfiguration(propertyDetails);
+			GetPropertyInheritanceConfiguration(
+				propertyDetails,
+				args,
+				hasValueInherits: out var hasValueInherits,
+				hasValueDoesNotInherit: out var hasValueDoesNotInherits,
+				valueAsProvider: out var newValueAsProvider
+			);
 
-			if (!hasValueDoesNotInherit && hasValueInherits)
+			if (!hasValueDoesNotInherits)
 			{
-				if (args.NewValue is IDependencyObjectStoreProvider provider)
+				if (hasValueInherits)
 				{
-					_childrenBindable[GetOrCreateChildBindablePropertyIndex(propertyDetails.Property)] =
-						provider.Store.Parent != ActualInstance ? args.NewValue : null;
+					if (newValueAsProvider != null)
+					{
+						_childrenBindable[GetOrCreateChildBindablePropertyIndex(propertyDetails.Property)]
+							= newValueAsProvider.Store.Parent != ActualInstance ? args.NewValue : null;
+					}
+					else
+					{
+						_childrenBindable[GetOrCreateChildBindablePropertyIndex(propertyDetails.Property)] = args.NewValue;
+					}
 				}
 				else
 				{
-					_childrenBindable[GetOrCreateChildBindablePropertyIndex(propertyDetails.Property)] = args.NewValue;
+					if (newValueAsProvider is { }
+						&& !(newValueAsProvider is UIElement))
+					{
+						_childrenBindable[GetOrCreateChildBindablePropertyIndex(propertyDetails.Property)]
+							= newValueAsProvider;
+					}
 				}
 			}
 		}
 
-		(bool hasValueInherits, bool hasValueDoesNotInherit) GetPropertyInheritanceConfiguration(DependencyPropertyDetails propertyDetails)
-		{
-			if (
-				propertyDetails.Property == _templatedParentProperty
-				|| propertyDetails.Property == _dataContextProperty
+		void GetPropertyInheritanceConfiguration(
+			DependencyPropertyDetails propertyDetails,
+			DependencyPropertyChangedEventArgs args,
+			out bool hasValueInherits,
+			out bool hasValueDoesNotInherit,
+			out IDependencyObjectStoreProvider? valueAsProvider
```

Comment:
I'm not really sure why this `out` parameter is needed? Seems like the `as` could be done just as easily in the calling method.

Reply:
The reason for the out is to avoid casting twice, once inside the method, and another one outside if either of the two boolean not true, but also because it removes the use of a generic ValueTuple.

Reply:
Reconsidering it, multiple alterations actually don't need this anymore. Will remove.